support-figma/extended-layout-plugin/src/design-spec-module.ts

```diff
@@ -0,0 +1,292 @@
+        topLevelComponents,
+        topLevelComponentDefns
+      );
+    }
+  }
+}
+
+// NOTE: this returns the id to component map.
+export async function mergeReactionsAsync(
+  topLevelComponents: Map<string, BaseNode[]>,
+  topLevelComponentDefns: Map<string, DesignComponentSpec>,
+  topLevelReactionComponents: Map<string, BaseNode[]>
+): Promise<Map<string, DesignComponentSpec>> {
+  for (const [
+    topLevelComponentNodeName,
+    topLevelComponentNodes,
+  ] of topLevelComponents) {
+    for (const node of topLevelComponentNodes) {
+      await populateReactionChainAsync(
+        node,
+        topLevelComponentNodeName,
+        topLevelReactionComponents
+      );
+    }
+  }
+
+  let idToDesignComponentSpecMap: Map<string, DesignComponentSpec> = new Map();
+  for (const [topLevelComponentNodeName, topLevelNodes] of topLevelComponents) {
+    let list = topLevelComponentDefns.get(topLevelComponentNodeName);
+    if (list) {
+      for (const topLevelNode of topLevelNodes) {
+        idToDesignComponentSpecMap.set(topLevelNode.id, list);
+      }
+    }
+  }
+  // Merge topLevelComponentDefns with topLevelReactionComponents
+  for (const [
+    topLevelComponentNodeName,
+    reactionNodes,
+  ] of topLevelReactionComponents) {
+    let list = topLevelComponentDefns.get(topLevelComponentNodeName);
+    if (list) {
+      for (const reactionNode of reactionNodes) {
+        idToDesignComponentSpecMap.set(reactionNode.id, list);
+      }
+    }
+  }
+  return idToDesignComponentSpecMap;
+}
+
+/** Take interaction into account. */
+async function populateReactionChainAsync(
+  node: BaseNode,
+  topLevelComponentNodeName: string,
+  topLevelReactionComponents: Map<string, BaseNode[]>
+) {
+  async function handleReactionNodeAsync(action?: Action) {
+    if (action?.type == "NODE") {
+      let destinationId = action?.destinationId;
+      if (destinationId) {
+        const destinationNode = await figma.getNodeByIdAsync(destinationId);
+        if (destinationNode) {
+          let list = topLevelReactionComponents.get(topLevelComponentNodeName);
+          if (list) {
+            if (!list.includes(destinationNode)) {
+              list.push(destinationNode);
+              await populateReactionChainAsync(
+                destinationNode,
+                topLevelComponentNodeName,
+                topLevelReactionComponents
+              );
+            }
+          } else {
+            topLevelReactionComponents.set(topLevelComponentNodeName, [
+              destinationNode,
+            ]);
+            await populateReactionChainAsync(
+              destinationNode,
+              topLevelComponentNodeName,
+              topLevelReactionComponents
+            );
+          }
+        }
+      }
+    }
+  }
+
+  var reactions = null;
+  let maybeReactions = node as ReactionMixin;
+  if (maybeReactions.reactions) {
+    reactions = maybeReactions.reactions;
+  }
+  if (reactions != null && reactions.length > 0) {
```

Comment:
Can you add some comments to describe what is going on here, and in the handleReactionNodAsync function above? Are you following all reaction types (navigate, change, overlay, etc) and adding their destination nodes to the list of nodes to look for text customizations?

Reply:
Done.